animation_nodes/data_structures/midi/midi_note.py

```diff
@@ -8,18 +8,34 @@ class MIDINote:
     timeOff: float = 0
     velocity: float = 0
 
-    def evaluate(self, time, attackTime, attackInterpolation, releaseTime, releaseInterpolation):
-        peakTime = self.timeOn + attackTime
-        endTime = self.timeOff + releaseTime
-        if self.timeOff >= time >= peakTime:
-            return 1
-        elif peakTime > time >= self.timeOn:
-            if attackTime == 0: return 1
-            return attackInterpolation((time - self.timeOn) / attackTime)
-        elif endTime >= time > self.timeOff:
-            if releaseTime == 0: return 1
-            return releaseInterpolation(1 - ((time - self.timeOff) / releaseTime))
-        return 0.0
+    def evaluate(self, time, attackTime, attackInterpolation, decayTime, decayInterpolation, sustainLevel, 
+        releaseTime, releaseInterpolation, velocitySensitivity):
+
+        def valueInEnvelope():
+            # find either point in time for envelope or where in envelope the timeOff happened
+            relativeTime = min(time, self.timeOff) - self.timeOn
+
+            if relativeTime <= 0.0:
+                return 0.0
+
+            if relativeTime < attackTime:
+                return attackInterpolation(relativeTime / attackTime)
+
+            relativeTime = relativeTime - attackTime
+
+            if relativeTime < decayTime:
+                decayNormalized = decayInterpolation(1 - relativeTime/ decayTime)
+                return decayNormalized * (1 - sustainLevel) + sustainLevel
+
+            return sustainLevel
+
+        value = valueInEnvelope()
+
+        if time > self.timeOff:
```

Comment:
I see no reason to put the release case outside like this. Since the values is guaranteed to be `sustainLevel`, just do `return sustainLevel * releaseInterpolation(...)` inside.

Reply:
Hmm, I don't think that works, because the time off could have taken place at any of the phases in the envelope. The function only returns the point in time in the envelope or the point that a timeOff took place.

Reply:
Right, my bad.